### ISSUE TITLE
Detecting step editor unmount

### ISSUE THREAD

**Issue by tskomudek** [nocode-js/sequential-workflow-designer#136]
Hi Bart,

when using a custom step editor in the react designer, i have the problem to loose the 
current changes in the form when the user accidentally clicks into the editor space
and the stepeditor component gets unmounted.

I remember you wrote somewhere that this is currently not possible in the react version,
but do you have an idea how to work around this ?

Thanks a lot

Tilo

**Comment by b4rtaz** [comment]
Hello @tskomudek,

basically you can always save the latest value in the step like:

```tsx
<input type="text" ... onChange={e => setProperty('x', e.value)} />
```

So in the definition you will have allways the latest changes.

If you have a save step between changing values and applying them, then you can create a hidden property in the step, such as `deltaX`.

```tsx
<input type="text" ... onChange={e => setProperty('deltaX', e.value)} />
```

During saving changes you need to move values to a final property (deltaX => x).

**Comment by b4rtaz** [comment]
@tskomudek have you resolved the problem?

**Comment by tskomudek** [comment]
Not really.
Yes, i can store the modified data somewhere, but i'd rather ask the user if he wants to navigate away from the current step without saving. Do you have any other idea how to achive this ?

Thx
Tilo

**Comment by b4rtaz** [comment]
It looks like you need a `canCloseEditor` callback or something like this. Yeah, it's not available now.

Maybe a temporary solution would be to use the `isSelectable` callback 🤔, something like:

```js
isSelectable: (step) => {
   if (hasUnsavedChanges()) return false;
   // ...
}
```

**Comment by ShraddhaFutrey** [comment]
When i am using sequential-workflow-designer in my angular app in router-outlet, its not taking full height and toolbox is not getting initialized.
i followed angular demo example

<img width="1270" alt="Screenshot 2024-05-24 at 4 32 24 PM" src="https://github.com/nocode-js/sequential-workflow-designer/assets/41083588/78066a1e-762c-4dfb-83ee-163282361e5f">
bb45)

**Comment by ShraddhaFutrey** [comment]
When i integrated sequential-workflow-designer-angular in my existing project in place of router outlet but its not getting initialised  properly...svg not taking full height and toolbox not getting initialized
![963066D4-C63D-4AE6-B37E-7D8205159105_1_201_a](https://github.com/nocode-js/sequential-workflow-designer/assets/41083588/a06abf15-cf63-4f68-b85f-372a0e05d3de)

**Comment by b4rtaz** [comment]
@ShraddhaFutrey it looks like the parent layers have not set any height constraints. The designer doesn't stretch parents, it adjusts to available space. You can check this theory by setting CSS: 

```css
.sqd-designer {height: 900px;}
```

BTW: please don't add comments to not related threads.

**Comment by tskomudek** [comment]
Hi Bart,

the isSelectable callback would help partially, but it doesn´t cover when the user just clicks into the canvas and the root editor appears or the user drags a new step from the toolbox.
An isUnselectable or canCloseEditor callback would be more suitable. Also, the onSelectedStepIdChanged event doesn't help here, since when it's called the stepeditor is already unmounted.
Maybe its worth considering at least.

Thanks
Tilo

**Comment by b4rtaz** [comment]
@tskomudek yes, this feature is planned but with a low priority. If you need this feature faster you can request it under the [consulting service](https://nocode-js.com/sequential-workflow-designer-pro-pricing#consulting).

**Comment by b4rtaz** [comment]
The feature is now available in version [0.33.0](https://github.com/nocode-js/sequential-workflow-designer/releases/tag/v0.33.0). You can now prevent unselection by implementing your own `canUnselectStep` callback. If unselection is blocked, the designer emits the `onStepUnselectionBlocked` event. You can find more details in the [documentation](https://nocode-js.com/docs/sequential-workflow-designer/features/editing-restrictions#can-unselect-step).

[Here’s an online demo](https://nocode-js.github.io/sequential-workflow-designer/react-app/#save-required-editor) showing how to implement a “save required” editor, when there are unsaved changes in the editor, the user cannot unselect the step. This demo is implemented in React ([here](https://github.com/nocode-js/sequential-workflow-designer/tree/main/demos/react-app/src/saveRequiredEditor) is the source code).
